### ISSUE TITLE
Training on multiple different length samples for Anomaly detection

### ISSUE THREAD

**Issue by yaseryacoob** [sintel-dev/Orion#279]
### Description

I am trying to train on a dataset of X samples of different temporal lengths of a process. The samples have equal step in time, but obviously can't be concatenated into one sample. What's the best way to train using Orion? 

I can think of
1. load the dataset at once, using a list of names (The documentation alludes to this)
1. iteratively fitting the model using the data something like
2   for all X train sequentially using orion.fit(train_data_i)

Any advice?

thank you

**Comment by sarahmish** [comment]
Hi @yaseryacoob! 

This is an interesting case. To understand your problem correctly, you have multiple dataframes of different lengths, do these dataframes represent the same sensor or are they different sensors?

If that's the case, you can go either approach, depending on the nature of the data
1. concatenate the data into a single dataframe `train_data`. This approach is recommended if the gap between one dataframe and another is small and they represent the same signal.
2. as for iteratively fitting on `tran_data_i`, this should work but is not supported by Orion.

To give you a more specific answer, would you mind answering the following:
* were the samples extracted from the sample signal?
* are the samples non-contiguous and hence cannot be concatenated?

**Comment by yaseryacoob** [comment]
Hi @sarahmish 

Thanks for responding.  Here are further clarifications:

1. Each time series is sampled at different experiment/occasion. The actual data is related to human behavior, the same person, but on different days, and can long time apart. As a result, concatenation "may work in general" if it weren't for the zone of concatenation (the end of one and the beginning of another) since it may introduce spurious features. But it should lead to modeling all the "variants" across the X samples. 
2. The sensors are different, but the features derived about the person behavior are robust to that, and therefore orion doesn't need to worry about that.
3. There is an important assumption, that because we are measuring the same person over and over, the variability is capped if we have enough samples.
4. Ideally, the training should be given a chance to model the different samples separately, so iterative minimization is more intuitive to me as the way to go (but may require modifications to some of the parameters (or more) so that a good model is reached.

Would do you mean by point 2. "not supported by orion"? Do you mean that orion.fit(tran_data_i) as in
```
for e in range(epoch_count):
   for i in range(len(X)):
        orion.fit(train_sample_i)
```
Will continuously fit the same model, or will recompute the fit from scratch in each call?

Thanks for your insights.

**Comment by sarahmish** [comment]
Thank you for the details @yaseryacoob!

Processing non-contiguous segments is one of the most [requested features](https://github.com/sintel-dev/Orion/issues/171), which is yet to be developed. We should push for this to happen!

To answer your question, currently (as of release 0.2.1), `tadgan` supports continuity, so your approach would work for that pipeline specifically. However other pipelines in Orion still don't. We are working on improving this in the next releases. See issue [here](https://github.com/MLBazaar/MLPrimitives/issues/267).
 
### How to use orion for non-contiguous segments
If you would like to use other pipelines, e.g. `lstm_dynamic_threshold`, I recommend preprocessing the data first to produce training windows, similar to this [primitive](https://sintel.dev/Orion/user_guides/primitives_pipelines/primitives/rolling_window_sequences.html).

There is also workaround to Orion to do this, which is applying a preprocessing pipeline to all the segments, then use another pipeline to train a the model.
For example, imagine you have the following pipeline `lstm_dynamic_threshold`, which has the following primitives:
```
"primitives": [
    "mlprimitives.custom.timeseries_preprocessing.time_segments_aggregate",
    "sklearn.impute.SimpleImputer",
    "sklearn.preprocessing.MinMaxScaler",
    "mlprimitives.custom.timeseries_preprocessing.rolling_window_sequences",
    "keras.Sequential.LSTMTimeSeriesRegressor",
    "orion.primitives.timeseries_errors.regression_errors",
    "orion.primitives.timeseries_anomalies.find_anomalies"
]
```
then you can breakdown the pipeline into two pipelines, a preprocessing pipeline and a modeling pipeline.

**preprocessing pipeline**
```
"primitives": [
    "mlprimitives.custom.timeseries_preprocessing.time_segments_aggregate",
    "sklearn.impute.SimpleImputer",
    "sklearn.preprocessing.MinMaxScaler",
    "mlprimitives.custom.timeseries_preprocessing.rolling_window_sequences"
]
```

**modeling pipeline**
```
"primitives": [
    "keras.Sequential.LSTMTimeSeriesRegressor",
    "orion.primitives.timeseries_errors.regression_errors",
    "orion.primitives.timeseries_anomalies.find_anomalies"
]
```

then you can apply the preprocessing to all the data, then concatenate them, and apply the modeling pipeline:
```python3
import pandas as pd

from mlblocks import MLPipeline
from orion import Orion

pipeline = MLPipeline("preprocessing.json")
pipeline.fit(train_data[0]) # fit on the first input (might need to adjust based on min-max scale)

processed_data = []
for train_data_i in train_data[1: ]:
    processed_data.append(pipeline.predict(train_data_i))

data = np.concatenate(processed_data)

orion = Orion("modeling.json")
orion.fit(data)
```

But again, this is not a clean way of approaching the problem, and we should support it internally in the primitive.

I hope this clarified a bit on the usage of Orion.

**Comment by yaseryacoob** [comment]
Many thanks @sarahmish for the detailed response. I have been using LSTM lstm_dynamic_threshold as I didn't see a clear advantage to using TADGAN. I will give TADGAN a try and compare performance. I am a little wary of chopping up data into training windows, and as you said it is not a clean way of approaching it, especially if one considers over-time modeling where data can flow in to "fine-tune" the model.  I

 am likely to need your advice again if I end up following this path, but you gave me a nice heads-up how to do it. Very much appreciated the detailed and thoughtful response!!!

**Comment by sarahmish** [comment]
Of course! I will close this issue for now, please open a new issue if you have any other questions!
